compiler/rustc_macros/src/symbols.rs

```diff
@@ -1,12 +1,373 @@
+//!
+//! * Finally, the next symbol value is assigned to the constant `DYNAMIC_SYMBOL_BASE`.
+//!
+//! # Performance benefits
+//!
+//! This design has several benefits:
+//!
+//! * When interning static symbols, there is no need to access thread-local storage or to
+//!   lock the interning table. This reduces thread contention on a global lock. Although
+//!   `rustc` does not currently rely on parallel compilation, it is a goal to move toward
+//!   parallel compilation. Preparing the symbol table now moves that goal closer.
+//!
+//!   Measurements show that, for many crates, the ratio of static symbols is between
+//!   45% and 75%. That is, depending on the crate, the number of symbol lookups that
+//!   turn out to be static lookups is typically 45% to 75%. This is a huge reduction
+//!   in lock contention.
+//!
+//! * Static symbols do not need to be duplicated into the heap. This reduces heap
+//!   allocations (no symbol copy, fewer hash table buckets). Since these symbols are
+//!   stored in shared pages, `rustc`'s working set is lower, at least during parsing.
+//!
+//! The main drawback of this design is the added complexity of the symbol lookups,
+//! both in this proc-macro and in the code that uses the static symbols. However,
+//! this complexity is well-encapsulated; the complexity is not visible to the users
+//! of `Symbol`.
+//!
+//! # Code generated by this proc-macro
+//!
+//! This proc-macro generates:
+//!
+//! * `kw::foo` constants for all of the keywords
+//! * `sym::foo` constants for all of the symbols
+//! * the `intern_static` function, which handles recognizing static symbols
+//! * constants which identify important ranges within the symbol value space,
+//!   such as `DYNAMIC_SYMBOL_BASE`.
+//!
+//! How we generate the  `intern_static` function deserves some explanation.
+//! It has a simple signature:
+//!
+//! ```text
+//! fn intern_static(string: &str) -> Option<Symbol> { ... }
+//! ```
+//!
+//! The contract is simple: if `string` is a static symbol, then `intern_static`
+//! returns `Some`. It does not use any dynamic data; it does not access any
+//! instance of `Interner`.
+//!
+//! The simplest, most naive implementation would be:
+//!
+//! ```text
+//! fn intern_static(string: &str) -> Option<Symbol> {
+//!     match string {
+//!         "if" => Some(kw::If),
+//!         "pub" => Some(kw::Pub),
+//!         // ...
+//!         _ => None,
+//!     }
+//! }
+//! ```
+//!
+//! Unfortunately, this translates to a sequential scan. Our goal is to generate
+//! a function which is considerably more efficient, but produces the same result.
+//!
+//! Our strategy is divide and conquer: We want to quickly eliminate strings that
+//! are not static strings, and for the remaining candidates, to partition potential
+//! results into smaller sets as quickly as possible. We have the advantage of
+//! being able to compute, at (proc-macro) compile time, information about the
+//! candidate strings.
```

Comment:
It would be nice if mir building could do this optimization. That would benefit everyone doing large matches on strings.

Reply:
Yes! I've written some prototype code for handling `match` over `&str` with lots of string-literal patterns. It would be great if `rustc` did this work, and then the special-case code in `symbols.rs` could be eliminated entirely.

If I can produce something reasonable, I'll work up a separate PR for that.